allennlp/data/iterators/homogeneous_batch_iterator.py

```diff
@@ -0,0 +1,74 @@
+from typing import Iterable, Dict, List
+import random
+from collections import defaultdict
+
+from allennlp.common.util import lazy_groups_of
+from allennlp.data.dataset import Batch
+from allennlp.data.instance import Instance
+from allennlp.data.iterators.data_iterator import DataIterator
+
+@DataIterator.register("homogeneous-batch")
+class HomogeneousBatchIterator(DataIterator):
+    """
+    This iterator takes a dataset of potentially heterogeneous instances
+    and yields back homogeneous batches. It assumes that each instance has
+    some ``MetadataField`` indicating what "type" of instance it is
+    and bases its notion of homogeneity on that (and, in particular, not on
+    inspecting the "field signature" of the instance.)
+
+    Parameters
+    ----------
+    batch_size : ``int``, optional, (default = 32)
+        The size of each batch of instances yielded when calling the iterator.
+    instances_per_epoch : ``int``, optional, (default = None)
+        If specified, each epoch will consist of precisely this many instances.
+        If not specified, each epoch will consist of a single pass through the dataset.
+    max_instances_in_memory : ``int``, optional, (default = None)
+        If specified, the iterator will load this many instances at a time into an
+        in-memory list and then produce batches from one such list at a time. This
+        could be useful if your instances are read lazily from disk.
+    cache_instances : ``bool``, optional, (default = False)
+        If true, the iterator will cache the tensorized instances in memory.
+        If false, it will do the tensorization anew each iteration.
+    track_epoch : ``bool``, optional, (default = False)
+        If true, each instance will get a ``MetadataField`` containing the epoch number.
+    partition_key : ``str``, optional, (default = "dataset")
```

Comment:
This is a bit gross, I wonder if it's better to allow setting an "origin" attribute on an `Instance` or something. Maybe not for this PR

Reply:
it needs to make it into the model, so if you did that you'd have to change all the "batch to tensor" logic to account for that and then make sure it doesn't somehow collide with other tensors and so on